src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs

```diff
@@ -18936,5 +18996,1565 @@ interface I1
+                //         if (h2 is S)
+                Diagnostic(ErrorCode.WRN_IsAlwaysTrue, "h2 is S").WithArguments("S").WithLocation(22, 13)
+                );
+
+            verifier.VerifyIL("Helper1<T>.Test1(T)",
+@"
+{ 
+  // Code size        7 (0x7)
+  .maxstack  1
+  IL_0000:  ldc.i4.1
+  IL_0001:  call       ""void System.Console.Write(int)""
+  IL_0006:  ret
+}
+");
+
+            verifier.VerifyIL("Helper2.Test2(S)",
+@"
+{
+  // Code size        7 (0x7)
+  .maxstack  1
+  IL_0000:  ldc.i4.3
+  IL_0001:  call       ""void System.Console.Write(int)""
+  IL_0006:  ret
+}
+");
+        }
+
+        [Fact]
+        public void IsOperator_05()
+        {
+            var src1 = @"
+class Helper1<T, U>
+    where T : allows ref struct
+{
+    public static void Test1(T h1)
+    {
+        if (h1 is U)
+        {
+        }
+    }
+}
+";
+
+            var comp = CreateCompilation(src1, targetFramework: s_targetFrameworkSupportingByRefLikeGenerics);
+            comp.VerifyDiagnostics(
+                // (7,13): error CS0019: Operator 'is' cannot be applied to operands of type 'T' and 'U'
+                //         if (h1 is U)
+                Diagnostic(ErrorCode.ERR_BadBinaryOps, "h1 is U").WithArguments("is", "T", "U").WithLocation(7, 13)
+                );
+
+            var src2 = @"
+class Helper2<U>
+{
+    public static void Test2(S h2)
+    {
+        if (h2 is U)
+        {
+            System.Console.Write(3);
+        }
+        else
+        {
+            System.Console.Write(4);
+        }
+    }
+}
+
+ref struct S : I1
+{
+}
+
+interface I1
+{
+}
+
+class Program
+{
+    static void Main()
+    {
+        Helper2<I1>.Test2(new S());
+    }
+}
+";
+            comp = CreateCompilation(src2, targetFramework: s_targetFrameworkSupportingByRefLikeGenerics, options: TestOptions.ReleaseExe);
+            var verifier = CompileAndVerify(
+                comp,
+                expectedOutput: ExecutionConditionUtil.IsMonoOrCoreClr ? "4" : null,
+                verify: ExecutionConditionUtil.IsMonoOrCoreClr ? Verification.Passes : Verification.Skipped).
+            VerifyDiagnostics(
+                // (6,13): warning CS0184: The given expression is never of the provided ('U') type
+                //         if (h2 is U)
+                Diagnostic(ErrorCode.WRN_IsAlwaysFalse, "h2 is U").WithArguments("U").WithLocation(6, 13)
+                );
+
+            verifier.VerifyIL("Helper2<U>.Test2(S)",
+@"
+{
+  // Code size        7 (0x7)
+  .maxstack  1
+  IL_0000:  ldc.i4.4
+  IL_0001:  call       ""void System.Console.Write(int)""
+  IL_0006:  ret
+}
+");
+        }
+
+        [Fact]
+        public void IsOperator_06()
+        {
+            var src = @"
+class Helper1<T, U>
+    where T : allows ref struct
+    where U : allows ref struct
+{
+    public static void Test1(T h1)
+    {
+        if (h1 is U)
+        {
+        }
+    }
+}
```

Comment:
Consider testing:
```csharp
class Helper1<T, U>
    where T : allows ref struct
    where U : T, allows ref struct
{
    public static void Test1(T h1)
    {
        if (h1 is U)
        {
        }
    }
    public static void Test2(U h2)
    {
        if (h2 is T)
        {
        }
    }
}
```

Reply:
> Consider testing:

I will add the following test in the next PR:
```
        [Fact]
        public void IsOperator_10()
        {
            var src = @"
class Helper1<T, U>
    where T : allows ref struct
    where U : T, allows ref struct
{
    public static void Test1(T h1)
    {
        if (h1 is U)
        {
        }
    }
    public static void Test2(U h2)
    {
        if (h2 is T)
        {
        }
    }
}
";

            var comp = CreateCompilation(src, targetFramework: s_targetFrameworkSupportingByRefLikeGenerics);
            comp.VerifyDiagnostics(
                // (8,13): error CS0019: Operator 'is' cannot be applied to operands of type 'T' and 'U'
                //         if (h1 is U)
                Diagnostic(ErrorCode.ERR_BadBinaryOps, "h1 is U").WithArguments("is", "T", "U").WithLocation(8, 13),
                // (14,13): error CS0019: Operator 'is' cannot be applied to operands of type 'U' and 'T'
                //         if (h2 is T)
                Diagnostic(ErrorCode.ERR_BadBinaryOps, "h2 is T").WithArguments("is", "U", "T").WithLocation(14, 13)
                );
        }
```